### ISSUE TITLE
Research spike on running in the background reliably

### ISSUE THREAD

**Issue by holmesworcester** [TryQuiet/quiet#916]
Goal: The Android app runs in the background well enough to send files, reliably receive notifications without a push notification service, and provide “liveness” to the network.

We should do some research to see how we can achieve these things. The end result of this ticket is a set of things we have proven do *not* work and a proposal for how to proceed. 

User story 1: “The Manager”: a manager wants to be able to provide quick responses to her team when they mention her. If the notification does not always arrive quickly, she cannot use Quiet. Quiet should always receive incoming notifications, without a push notification service.
- If someone sends a message to a channel and notifications are enabled, she should see a notification within 3 seconds.
- When she restarts her phone she should see notifications, whether or not she opens Quiet. These notifications should appear within seconds or, if not possible, minutes.
- Whenever Quiet is not able to receive notifications for some reason, she should see some indication of this, e.g. a notification saying “Quiet not connected–reconnect?” 

User story 2: “The Always-on Phone”: a team wants to make sure there is always a place to sync messages, but they don’t have a spare computer to use, so they join the network on an Android phone and leave that phone plugged in 24/7. That phone should remain reliably connected to the network.
- There should be something the user can do to ensure Quiet will stay connected 24/7 for multiple days, syncing messages.
- Phone stays connected for 7 days (indefinitely)
- This works on any phone that runs Quiet

User story 3: “Sending an image”: a user takes a photo, sends it with Quiet to others in their team, and puts their phone in their pocket. The photo is important and must be delivered to all online users.
- The photo is always delivered
- The app is not killed while the photo is being delivered

**Comment by holmesworcester** [comment]
Relevant docs:
1. https://developer.android.com/guide/components/foreground-services
2. https://developer.android.com/reference/androidx/work/Worker
3. https://developer.android.com/reference/androidx/work/WorkManager
4. Anything else...?

**Comment by siepra** [comment]
see https://github.com/TryQuiet/ZbayMobile/issues/93

**Comment by holmesworcester** [comment]
Our ideas for things to look into:

- Workers that run periodically (for #1 and #4)
- Foreground services that monitor and tell user when they are offline (for #1 and #3)
- Foreground services in their most aggressive form that keep our threads alive somehow if the phone is charging and not being used for anything else (for #2, maybe?)

**Comment by siepra** [comment]
I think we should start with getting more information and control over our Threads - I've got some ideas on how to achieve that but it requires some time and native coding

**Comment by siepra** [comment]
I set up workers which runs every 15 minutes (starting in 10 minutes from the app start). They display information about backend process status. If data port is no longer occupied, worker should trigger a push notification informing that app no longer works. I think I did not receive this notification on my phone but maybe I just didn't pay enough attention

**Comment by siepra** [comment]
We want to confirm workers works in a desired way (they are capable of performing some work when backend thread is dead) and then find a way to gather similar information about tor process status within the same worker

**Comment by siepra** [comment]
Notifications triggered from worker may need access to UI thread as they appear randomly when the app is being re-started or phone being unlocked. It'd be good to check that by reading documentation... 

Also I received the notification but app seemed to keep working so maybe checking data server port occupancy is not good enough

**Comment by siepra** [comment]
I received the notification but app seemed to keep working so maybe checking data server port occupancy is not good enough

**Comment by siepra** [comment]
I think the good idea may be to dump workers logs into a file and gather them for some time to have the overall picture of what happens to the app and it's threads

**Comment by siepra** [comment]
I added dumping worker logs to a file. To upload this file from device to local file system run `adb -d shell "run-as com.zbaymobile.debug cat /data/data/com.zbaymobile.debug/files/worker_logs.txt" > Desktop/worker_logs.txt`

**Comment by siepra** [comment]
I was trying to gather some data picturing real workers behavior but those are not very consistent (I still don't get it why some scheduled workers seems to trigger only after waking up the phone/app and why notifications are being displayed randomly).
This is what I've tried:
- Logging directly from worker (couldn't get consistent data)
- Logging scheduled work requests statuses (never reached)
- Using inspection tools from Android Studio (all workers appearing as 'enqueued')

**Comment by siepra** [comment]
Workers seems to run reliably every 15 minutes but the question is **are there situations where workers run but backend and tor are dead?**

**Comment by siepra** [comment]
I've been testing workers during the night and those are my observations:
- Workers worked reliably (I collected all of the logs - 4 logs per hour for 17 hours)
- Data server closed around 2 am - workers remained alive
- Sometimes I didn't know if I'm not connected to anyone or if the app is broken (missing network state info)

**Comment by holmesworcester** [comment]
Can we run backend in workers? Or start it with workers?

**Comment by siepra** [comment]
Note: we treat user story 4 as a separate issue https://github.com/TryQuiet/quiet/issues/985
